emulator_test.go

```diff
@@ -262,8 +265,122 @@ func TestGetQueuePreviouslyExisted(t *testing.T) {
+		Task: &taskspb.Task{
+			Name: createdQueue.GetName() + "/tasks/any-task",
+			MessageType: &taskspb.Task_HttpRequest{
+				HttpRequest: &taskspb.HttpRequest{
+					// Use the not_found handler to prove that purge stops any further retries
+					Url: "http://localhost:5000/not_found",
+				},
+			},
+		},
+	}
+	createdTask, err := client.CreateTask(context.Background(), &createTaskRequest)
+	require.NoError(t, err)
+
+	// Task was created OK, verify that the first HTTP request was sent
+	_, err = awaitHttpRequest(receivedRequests)
+	require.NoError(t, err)
+
+	// Now purge the queue
+	purgeQueueRequest := taskspb.PurgeQueueRequest{
+		Name: createdQueue.GetName(),
+	}
+	_, err = client.PurgeQueue(context.Background(), &purgeQueueRequest)
+	require.NoError(t, err)
+
+	// Wait a moment for that to work, then verify nothing in the list and cannot retrieve by name
+	time.Sleep(100 * time.Millisecond)
+	assertTaskListIsEmpty(t, client, createdQueue)
+	assertGetTaskFails(t, grpcCodes.FailedPrecondition, client, createdTask.GetName())
+
+	// BUT - Verify that the task name is still not available for new tasks
+	_, err = client.CreateTask(context.Background(), &createTaskRequest)
+	if assert.Error(t, err, "Should still prevent duplicate names after purging queue") {
+		rsp, _ := grpcStatus.FromError(err)
+		assert.Equal(t, grpcCodes.AlreadyExists, rsp.Code())
+	}
+
+	// Verify that it only sent the original HTTP request, it purged before the retries
+	_, err = awaitHttpRequestWithTimeout(receivedRequests, 1*time.Second)
+	assert.Error(t, err, "Should not receive any further HTTP requests within timeout")
+}
+
+func TestPurgeQueueOptionallyPerformsHardReset(t *testing.T) {
+	serv, client := setUp(t, ServerOptions{HardResetOnPurgeQueue: true})
+	defer tearDown(t, serv)
+
+	createdQueue := createTestQueue(t, client)
+	defer tearDownQueue(t, client, createdQueue)
+
+	srv, receivedRequests := startTestServer()
+	defer srv.Shutdown(context.Background())
+
+	createTaskRequest := taskspb.CreateTaskRequest{
+		Parent: createdQueue.GetName(),
+		Task: &taskspb.Task{
+			Name: createdQueue.GetName() + "/tasks/any-task",
+			MessageType: &taskspb.Task_HttpRequest{
+				HttpRequest: &taskspb.HttpRequest{
+					// Use the not_found handler to prove that purge stops any further retries
+					Url: "http://localhost:5000/not_found",
+				},
+			},
+		},
+	}
+	createdTask, err := client.CreateTask(context.Background(), &createTaskRequest)
+	require.NoError(t, err)
+
+	// Task was created OK, verify that the first HTTP request was sent
+	_, err = awaitHttpRequest(receivedRequests)
+	require.NoError(t, err)
+
+	// Now purge the queue
+	purgeQueueRequest := taskspb.PurgeQueueRequest{
+		Name: createdQueue.GetName(),
+	}
+	_, err = client.PurgeQueue(context.Background(), &purgeQueueRequest)
+	require.NoError(t, err)
+
+	// In this mode, purging the queue is synchronous so we should be in the empty state straight away
+	log.Printf("Purge Queue returned")
```

Comment:
Did you intend to leave this?

Reply:
Whoops - no! Will remove when I update this branch.

Reply:
Removed this line within the rebased commits.